### ISSUE TITLE
Data API - Add a non-changing unique identifier

### ISSUE THREAD

**Issue by 949mac** [2020PB/police-brutality#236]
Hello - For the Data API, can you add a non-changing ID to each record? Such as the PR# or any other unique id?

I'm writing a script that is analyzing the links, which comes at an expensive cost. I would like the links to only be analyzed once. It would be helpful if each record had some internal ID we could reference when updating or aggregating additional data.

For example:

```
data: [
{
   id: 1, // <-- Add this, keep it consistent across all data api updates ❗
   links: [
      "https://twitter.com/courtenay_roche/status/1267653137969623040",
      "https://twitter.com/yagirlbrookie09/status/1267647898365427714",
      "https://www.4029tv.com/article/bentonville-police-deploy-tear-gas-on-protesters/32736629#"
   ],
   state: "Arkansas",
   edit_at: "https://github.com/2020PB/police-brutality/blob/master/reports/Arkansas.md",
   city: "Bentonville",
   name: "Law enforcement gas a crowd chanting “we want peace” right after exiting the building.",
   date: "2020-06-01",
   date_text: "June 1st"
},
```

Much appreciated. Thanks for the wonderful contribution.

**Comment by ubershmekel** [comment]
Thank you for this issue. I was just thinking about it too. Because the source of truth is the markdown - there are two options I see:

* A github action script that gives each issue an id. I dislike having bots editing the raw files.
* We create an id that will be unique, and editors can write in the markdown.

Here's an example:

```
## Compton

### Police beat person on the ground | (believed to be) June 1st

Two police offers have someone pinned to the ground and proceed to kneel on their head/neck, punch and kick them. A third officer joins them, also striking the person.

[This Reddit comment](https://www.reddit.com/r/PublicFreakout/comments/gvdl01/they_secluded_him_behind_a_wall_and_looked_around/fsolupz/) asserts these police are members of the LASD.

**Links**

* https://www.reddit.com/r/PublicFreakout/comments/gvdl01/they_secluded_him_behind_a_wall_and_looked_around/

* Incident id: ca-2020-06-01-secluded
```

That format is `<state-abbreviation>-<iso-date>-<letters>`. What do you think? Would love to hear feedback or a better idea.

**Comment by 949mac** [comment]
Hi @ubershmekel,

Why not just use the PR ID? That way it can be traced back to the source. That's the easiest in my mind... but truth be told, any unique id is better than none. Maybe a combination of your suggestion w/ PR ID?

`<state-abbreviation>-<iso-date>-<letters>-<pr-id>` 

In this case, the PR id is still present for reference purposes to trace the object.

**Comment by ubershmekel** [comment]
I'm imagining a random person coming in, hitting the github edit button, and trying to figure out what that id should be. I don't think they'd know the PR id when they're editing.

We can always `blame` to get the original submission.

**Comment by 949mac** [comment]
@ubershmekel -- it should be super easy for the end-user adding the ID. In this case, we'll just need to worry about collisions (duplicate IDs). Realistically this is a problem for thousands of records. If this is hundreds of records, it might not be as risky.

**Comment by 949mac** [comment]
Sorry -- I just had an idea. If you incorporate the person's username, initials, etc, then you can do:

`20200604-2114-949mac`

That is...

`<date in hhmmdd>-<24hrtime>-<username>`

This will give you a unique username for every generation without the need to check any external sources, regardless of timezone.

Just a thought.

**Comment by bonedaddy** [comment]
> @ubershmekel -- it should be super easy for the end-user adding the ID. In this case, we'll just need to worry about collisions (duplicate IDs). Realistically this is a problem for thousands of records. If this is hundreds of records, it might not be as risky.

Using the ID of a PR seems like the easiest option, as its pretty simple for anyone adding a new PR to do. If a UUID is being used, [ksuid](https://github.com/segmentio/ksuid) is pretty cool, as it can be sorted based on time.

**Comment by 949mac** [comment]
> > @ubershmekel -- it should be super easy for the end-user adding the ID. In this case, we'll just need to worry about collisions (duplicate IDs). Realistically this is a problem for thousands of records. If this is hundreds of records, it might not be as risky.
> 
> Using the ID of a PR seems like the easiest option, as its pretty simple for anyone adding a new PR to do. If a UUID is being used, [ksuid](https://github.com/segmentio/ksuid) is pretty cool, as it can be sorted based on time.

Using a UUID would actually my preferred option. +1 on that

**Comment by 01210210** [comment]
Doesnt using PR id assume that only one incident is ever added per PR? I didn't see that in the contribution guidelines and it appears people are updating states daily.

**Comment by 949mac** [comment]
> Doesnt using PR id assume that only one incident is ever added per PR? I didn't see that in the contribution guidelines and it appears people are updating states daily.

That's a good point. I think UUIDs would be best if the contributors could find a consistent and easy way to add it.

**Comment by colegleason** [comment]
Could the new issue template include a link to https://www.uuidgenerator.net/ and ask the contributor to paste one in?

**Comment by ubershmekel** [comment]
I'll write a script that will give the existing incidents a uuid

**Comment by ubershmekel** [comment]
Actually, the thing is, I worry that these ids will be used in URLs and that would make these URLs very long and less readable. So I'll make the script do something like <state-abbreviated>-<city-abbreviated>-<num>

**Comment by andrewsuzuki** [comment]
After #324, it seems eight events are still missing ids.

**Comment by cwilper** [comment]
I'm getting 11 missing ids currently. @ubershmekel Here is a script to identify records (from the json) which are current missing ids: https://gist.github.com/cwilper/2e192e3ea15ca689c4d891f4f9c77c26

Happy to submit a PR to fix them, but should the records be re-ordered to that they are sequential as part of that?
